app/main.py

```diff
@@ -1,3 +1,66 @@
+from typing import Dict, Any
+
+
+def apply_armour(knight: Dict[str, Any]) -> int:
+    """Calculate total armour protection for a knight."""
+    return sum(piece["protection"] for piece in knight["armour"])
+
+
+def apply_weapon(knight: Dict[str, Any]) -> int:
+    """Calculate total power with the weapon applied."""
+    return knight["power"] + knight["weapon"]["power"]
+
+
+def apply_potion(knight: Dict[str, Any]) -> None:
+    """Apply potion effects to the knight, if any."""
+    if knight["potion"]:
+        effect = knight["potion"]["effect"]
+        knight["power"] += effect.get("power", 0)
+        knight["protection"] += effect.get("protection", 0)
+        knight["hp"] += effect.get("hp", 0)
+
+
+def prepare_knight(knight: Dict[str, Any]) -> Dict[str, Any]:
+    """Prepare knight with armour, weapon, and potion."""
+    knight["protection"] = apply_armour(knight)
```

Comment:
The function `apply_armour` assumes that the 'armour' key is always present and is a list. Consider adding a check to ensure 'armour' is present and is a list to avoid potential errors.